File: frontEnd/src/stores/interview.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

// ── Types ──────────────────────────────────────────────────

export interface JobPosition {
  id: string
  title: string
  icon: string
  description: string
}

export interface JobCategory {
  category: string
  positions: JobPosition[]
}

export interface RoundProgress {
  round_key: string
  label: string
  status: 'pending' | 'active' | 'completed'
}

export interface InterviewSession {
  id: string
  job_category: string
  job_title: string
  current_round: string
  status: string
  cheat_count: number
  interview_mode: string
  target_round: string | null
  rounds_progress: RoundProgress[]
  started_at: string
}

export interface QuestionItem {
  id: string
  question_type: 'choice' | 'judgment' | 'code' | 'open_ended'
  content: Record<string, unknown>
  time_limit: number
}

export interface AnswerResponse {
  correct: boolean
  score: number
  feedback: string
  correct_answer: string | Record<string, unknown> | null
}

export interface AIChatMessage {
  role: 'interviewer' | 'candidate'
  content: string
}

export interface RadarData {
  professional: number
  logic: number
  communication: number
  match: number
}

export interface RoundDetail {
  round_key: string
  label: string
  score: number
  max_score: number
  answers: Record<string, unknown>[]
}

export interface InterviewReport {
  session_id: string
  job_category: string
  job_title: string
  total_score: number
  max_total: number
  grade: string
  radar: RadarData
  round_details: RoundDetail[]
  suggestions: string[]
  ai_analysis: string
  completed_at: string | null
  interview_mode: string
  target_round: string | null
}

export interface InterviewHistoryItem {
  id: string
  job_category: string
  job_title: string
  status: string
  total_score: number | null
  grade: string | null
  cheat_count: number
  interview_mode: string
  target_round: string | null
  started_at: string
  completed_at: string | null
}

// ── API Client ─────────────────────────────────────────────

const API_BASE = '/api'

function getAuthHeaders(): Record<string, string> {
  const token = localStorage.getItem('auth_token')
  return {
    'Content-Type': 'application/json',
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
  }
}

async function apiRequest<T>(path: string, options: RequestInit = {}): Promise<T> {
  const headers = getAuthHeaders()
  const res = await fetch(`${API_BASE}${path}`, {
    ...options,
    headers: { ...headers, ...(options.headers as Record<string, string> || {}) },
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw new Error(body.detail || `请求失败 (${res.status})`)
  }
  return res.json()
}

// ── Store ──────────────────────────────────────────────────

export const useInterviewStore = defineStore('interview', () => {
  const jobCategories = ref<JobCategory[]>([])
  const currentSession = ref<InterviewSession | null>(null)
  const currentQuestions = ref<QuestionItem[]>([])
  const aiMessages = ref<AIChatMessage[]>([])
  const report = ref<InterviewReport | null>(null)
  const historyItems = ref<InterviewHistoryItem[]>([])
  const loading = ref(false)

  const isActive = computed(() => currentSession.value?.status === 'in_progress')
  const currentRound = computed(() => currentSession.value?.current_round || '')

  async function fetchJobs() {
    loading.value = true
    try {
      jobCategories.value = await apiRequest<JobCategory[]>('/interview/jobs')
    } finally {
      loading.value = false
    }
  }

  async function startInterview(
    jobCategory: string,
    jobTitle: string,
    interviewMode: string = 'full',
    targetRound?: string | null,
  ): Promise<InterviewSession> {
    loading.value = true
    try {
      const session = await apiRequest<InterviewSession>('/interview/start', {
        method: 'POST',
        body: JSON.stringify({
          job_category: jobCategory,
          job_title: jobTitle,
          interview_mode: interviewMode,
          target_round: targetRound || null,
        }),
      })
      currentSession.value = session
      return session
    } finally {
      loading.value = false
    }
  }

  async function fetchSession(sessionId: string) {
    currentSession.value = await apiRequest<InterviewSession>(`/interview/session/${sessionId}`)
  }

  async function fetchQuestions(sessionId: string): Promise<QuestionItem[]> {
    const data = await apiRequest<{ round: string; questions: QuestionItem[] }>(
      `/interview/session/${sessionId}/question`
    )
    currentQuestions.value = data.questions
    return data.questions
  }

  async function submitAnswer(
    sessionId: string,
    questionId: string,
    answer: string | Record<string, unknown>,
    durationSeconds: number
  ): Promise<AnswerResponse> {
    return apiRequest<AnswerResponse>(`/interview/session/${sessionId}/answer`, {
      method: 'POST',
      body: JSON.stringify({
        question_id: questionId,
        answer,
        duration_seconds: durationSeconds,
      }),
    })
  }

  async function nextRound(sessionId: string): Promise<InterviewSession> {
    const session = await apiRequest<InterviewSession>(`/interview/session/${sessionId}/next`, {
      method: 'POST',
    })
    currentSession.value = session
    return session
  }

  async function sendAIChat(
    sessionId: string,
    messages: { role: string; content: string }[],
    round: string,
    onChunk: (text: string) => void,
    signal?: AbortSignal,
  ): Promise<string> {
    const token = localStorage.getItem('auth_token')
    const res = await fetch(`${API_BASE}/interview/session/${sessionId}/ai-chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      body: JSON.stringify({ messages, round }),
      signal,
    })
    if (!res.ok) {
      throw new Error(`AI对话请求失败 (${res.status})`)
    }

    const reader = res.body?.getReader()
    const decoder = new TextDecoder()
    let fullText = ''

    if (reader) {
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        const text = decoder.decode(value, { stream: true })
        const lines = text.split('\n')
        for (const line of lines) {
          if (!line.startsWith('data: ')) continue
          const dataStr = line.slice(6).trim()
          if (dataStr === '[DONE]') break
          try {
            const parsed = JSON.parse(dataStr)
            if (parsed.content) {
              fullText += parsed.content
              onChunk(fullText)
            }
          } catch { /* skip */ }
        }
      }
    }
    return fullText
  }

  async function reportCheat(sessionId: string, cheatCount: number) {
    const session = await apiRequest<InterviewSession>(`/interview/session/${sessionId}/cheat`, {
      method: 'POST',
      body: JSON.stringify({ cheat_count: cheatCount }),
    })
    currentSession.value = session
  }

  async function abortInterview(sessionId: string) {
    const session = await apiRequest<InterviewSession>(`/interview/session/${sessionId}/abort`, {
      method: 'POST',
    })
    currentSession.value = session
  }

  async function fetchReport(sessionId: string): Promise<InterviewReport> {
    report.value = await apiRequest<InterviewReport>(`/interview/session/${sessionId}/report`)
    return report.value
  }

  async function fetchHistory() {
    const data = await apiRequest<{ total: number; sessions: InterviewHistoryItem[] }>(
      '/interview/history'
    )
    historyItems.value = data.sessions
  }

  function resetSession() {
    currentSession.value = null
    currentQuestions.value = []
    aiMessages.value = []
    report.value = null
  }

  return {
    jobCategories,
    currentSession,
    currentQuestions,
    aiMessages,
    report,
    historyItems,
    loading,
    isActive,
    currentRound,
    fetchJobs,
    startInterview,
    fetchSession,
    fetchQuestions,
    submitAnswer,
    nextRound,
    sendAIChat,
    reportCheat,
    abortInterview,
    fetchReport,
    fetchHistory,
    resetSession,
  }
})

File: backEnd/app/schemas/interview.py
from datetime import datetime
from pydantic import BaseModel, Field


# ---------- 岗位列表 ----------

class JobPosition(BaseModel):
    id: str
    title: str
    icon: str = ""
    description: str = ""


class JobCategory(BaseModel):
    category: str
    positions: list[JobPosition]


# ---------- 面试会话 ----------

class RoundProgress(BaseModel):
    round_key: str  # assessment / tech / business / ai_voice_3 / ai_voice_4
    label: str
    status: str  # pending / active / completed


class InterviewSessionCreate(BaseModel):
    job_category: str = Field(..., min_length=1)
    job_title: str = Field(..., min_length=1)
    interview_mode: str = "full"  # full / single
    target_round: str | None = None  # 仅 single 模式


class InterviewSessionResponse(BaseModel):
    id: str
    job_category: str
    job_title: str
    current_round: str
    status: str
    cheat_count: int
    interview_mode: str = "full"
    target_round: str | None = None
    rounds_progress: list[RoundProgress]
    started_at: datetime

    model_config = {"from_attributes": True}


# ---------- 题目 ----------

class QuestionItem(BaseModel):
    id: str
    question_type: str  # choice / judgment / code / open_ended
    content: dict
    time_limit: int = 30  # 秒


class QuestionListResponse(BaseModel):
    round: str
    questions: list[QuestionItem]


# ---------- 答题 ----------

class AnswerSubmit(BaseModel):
    question_id: str = Field(..., min_length=1)
    answer: str | dict  # 选择题选项, 开放题文本, 代码题 code+language
    duration_seconds: int = 0


class AnswerResponse(BaseModel):
    correct: bool
    score: float
    feedback: str
    correct_answer: str | dict | None = None


# ---------- AI 对话 ----------

class AIChatRequest(BaseModel):
    messages: list[dict]  # [{ role, content }]
    round: str  # ai_voice_3 / ai_voice_4


class AIChatMessage(BaseModel):
    role: str
    content: str


# ---------- 切屏上报 ----------

class CheatReport(BaseModel):
    cheat_count: int


# ---------- 评分报告 ----------

class RoundDetail(BaseModel):
    round_key: str
    label: str
    score: float
    max_score: float
    answers: list[dict]  # 每题 { question, user_answer, correct, feedback }


class RadarData(BaseModel):
    professional: float  # 专业能力
    logic: float  # 逻辑思维
    communication: float  # 沟通表达
    match: float  # 岗位匹配度


class InterviewReport(BaseModel):
    session_id: str
    job_category: str
    job_title: str
    interview_mode: str = "full"
    target_round: str | None = None
    total_score: float
    max_total: float
    grade: str  # A / B / C / D
    radar: RadarData
    round_details: list[RoundDetail]
    suggestions: list[str]
    ai_analysis: str = ""
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


# ---------- 历史记录 ----------

class InterviewHistoryItem(BaseModel):
    id: str
    job_category: str
    job_title: str
    status: str
    total_score: float | None
    grade: str | None
    cheat_count: int
    interview_mode: str
    target_round: str | None
    started_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class InterviewHistoryResponse(BaseModel):
    total: int
    sessions: list[InterviewHistoryItem]
